internal/state/relationship/capturer.go

```diff
@@ -0,0 +1,131 @@
+// The relationships between HTTPRoutes -> Services are many to 1, so these relationships are tracked using a counter. A Service relationship exists if at least one HTTPRoute references it.
+// An EndpointSlice relationship exists, if its Service owner is referenced by at least one HTTPRoute.
+type Capturer struct {
+	routesToServices    map[types.NamespacedName]map[types.NamespacedName]struct{}
+	serviceRefCount     map[types.NamespacedName]int
+	endpointSliceOwners map[types.NamespacedName]types.NamespacedName
+}
+
+// NewCapturer creates a new instance of Capturer.
+func NewCapturer() *Capturer {
+	return &Capturer{
+		routesToServices:    make(map[types.NamespacedName]map[types.NamespacedName]struct{}),
+		serviceRefCount:     make(map[types.NamespacedName]int),
+		endpointSliceOwners: make(map[types.NamespacedName]types.NamespacedName),
+	}
+}
+
+// Capture captures relationships for the given object.
+func (c *Capturer) Capture(obj client.Object) {
+	switch o := obj.(type) {
+	case *v1beta1.HTTPRoute:
+		c.upsertForRoute(o)
+	case *discoveryV1.EndpointSlice:
+		svcName := sdk.GetServiceNameFromEndpointSlice(o)
+		if svcName != "" {
+			c.endpointSliceOwners[client.ObjectKeyFromObject(o)] = types.NamespacedName{Namespace: o.Namespace, Name: svcName}
+		}
+	}
+}
+
+// Remove removes the relationship for the given object from the Capturer.
+func (c *Capturer) Remove(resourceType client.Object, nsname types.NamespacedName) {
+	switch resourceType.(type) {
+	case *v1beta1.HTTPRoute:
+		c.deleteForRoute(nsname)
+	case *discoveryV1.EndpointSlice:
+		delete(c.endpointSliceOwners, nsname)
+	}
+}
+
+// Exists returns true if the given object has a relationship with another object.
+func (c *Capturer) Exists(resourceType client.Object, nsname types.NamespacedName) bool {
+	switch resourceType.(type) {
+	case *v1.Service:
+		return c.serviceRefCount[nsname] > 0
+	case *discoveryV1.EndpointSlice:
+		svcOwner, exists := c.endpointSliceOwners[nsname]
+		return exists && c.serviceRefCount[svcOwner] > 0
+	}
+
+	return false
+}
+
+func (c *Capturer) upsertForRoute(route *v1beta1.HTTPRoute) {
+	oldServices := c.routesToServices[client.ObjectKeyFromObject(route)]
+	newServices := getBackendServiceNamesFromRoute(route)
+
+	for svc := range oldServices {
+		if _, exist := newServices[svc]; !exist {
+			c.decrementRefCount(svc)
+		}
+	}
+
+	for svc := range newServices {
+		if _, exist := oldServices[svc]; !exist {
+			c.serviceRefCount[svc]++
+		}
+	}
+
+	c.routesToServices[client.ObjectKeyFromObject(route)] = newServices
+}
+
+func (c *Capturer) deleteForRoute(routeName types.NamespacedName) {
+	services := c.routesToServices[routeName]
+
+	for svc := range services {
+		c.decrementRefCount(svc)
+	}
+
+	delete(c.routesToServices, routeName)
+}
+
+func (c *Capturer) decrementRefCount(svcName types.NamespacedName) {
```

Comment:
This function can leave us with a key and -1.

```
package main

import "fmt"

func dec(m map[string]int, key string) {
	if m[key] == 1 {
		delete(m, key)
		return
	}
	m[key]--
}

func main() {
	m := make(map[string]int)

	dec(m, "hello")
	fmt.Printf("%+v\n", m)
}
```

Reply:
`map[hello:-1]`

Reply:
We may need a test that exercises key existence versus non-existence.

Reply:
Thanks for catching this.

I fixed it and am now explicitly testing the ref count.